### PR TITLE
Vine: Thread Controller for FunctionCalls

### PR DESCRIPTION
**Pull request by JinZhou5042**
## Proposed changes

Fix for #3734 

A lightweight thread resource controller for FunctionCall tasks, adopted from [threadpoolctl](https://github.com/joblib/threadpoolctl) but more compact, with the minimum overhead to manage thread numbers at runtime.

## Post-change actions

Put an 'x' in the boxes that describe post-change actions that you have done.
The more 'x' ticked, the faster your changes are accepted by maintainers.

- [ ] `make test`       Run local tests prior to pushing.
- [ ] `make format`     Format source code to comply with lint policies. Note that some lint errors can only be resolved manually (e.g., Python)
- [ ] `make lint`       Run lint on source code prior to pushing.
- [ ] Manual Update     Did you update the manual to reflect your changes, if appropriate? This action should be done after your changes are approved but not merged.
- [ ] Type Labels       Select github labels for the type of this change: bug, enhancement, etc.
- [ ] Product Labels    Select github labels for the product affected: TaskVine, Makeflow, etc.
- [ ] PR RTM            Mark your PR as ready to merge.

## Additional comments
This section is dedicated to changes that are ambitious or complex and require substantial discussions. Feel free to start the ball rolling.

### CODE REVIEW

File: poncho/src/poncho/library_network_code.py [hunk 1]
```diff
@@ -20,6 +19,128 @@ def library_network_code():
     import cloudpickle
     import select
     import signal
+    import ctypes
+    from ctypes.util import find_library
+    from contextlib import ContextDecorator
+
+    class CLibController:
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:26]: Add comment to give attribution to original implementation.

File: poncho/src/poncho/library_network_code.py [hunk 2]
```diff
@@ -106,50 +227,59 @@ def start_function(in_pipe_fd):
...
-                    with open('outfile', 'wb') as f:
-                        cloudpickle.dump(globals()[function_name](event), f)
-                    os.close(stdout_capture_fd)
-                    os._exit(0)
-                elif p < 0:
-                    print(f'Library code: unable to fork to execute {function_name}', file=sys.stderr)
-                    return -1
-
-                # return pid and function id of child process to parent.
+            with ThreadpoolLimits(limits=library_cores // function_slots):
```
Comment (btovar) [inline poncho/src/poncho/library_network_code.py:230]: Add comment to give attribution to original implementation.

**Review by btovar — CHANGES_REQUESTED** [review]

**Comment by JinZhou5042** [comment]
The adopted little thread pool controller involves more variations of mathematical packages on the C site compared to `threadpoolctl`, including _flexiblas_, _mkl_core_, _libmkl_rt_, _gomp_, _iomp5_, _libiomp5_, _libomp_ and _vcomp140_, which may prevent core over-commitment under some unexpected conditions. And we can simply update it by inserting a couple of lines once detecting some un-configured libraries. 

However, directly importing the original package is easier and it makes the library code more compact. I think either is fine, how do you think @btovar

**Comment by dthain** [comment]
@colinthomas-z80 what are you thoughts on how this affects the Parsl install?  While we strongly encourage Conda installs, I believe that their CI builds download our tarball and then do some `pip`.

**Comment by colinthomas-z80** [comment]
@dthain I'm not sure what the issue is exactly? these new imports are part of the standard library I believe

**Pull request closed without merging**